src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs

```diff
@@ -360,44 +424,17 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
 
             if (propertyFilter != null)
             {
-                object subject = _page;
-
-                if (_model != null)
-                {
-                    subject = _model;
-                }
-
-                propertyFilter.Subject = subject;
+                propertyFilter.Subject = _model;
                 propertyFilter.ApplyTempDataChanges(_pageContext.HttpContext);
             }
 
-            IActionResult result = null;
-
-            var handler = _selector.Select(_pageContext);
-            if (handler != null)
+            _result = await ExecuteHandlerMethod(_model);
+            if (_result is PageResult pageResult)
             {
-                var arguments = await GetArguments(handler);
-
-                Func<object, object[], Task<IActionResult>> executor = null;
-                for (var i = 0; i < actionDescriptor.HandlerMethods.Count; i++)
-                {
-                    if (object.ReferenceEquals(handler, actionDescriptor.HandlerMethods[i]))
-                    {
-                        executor = CacheEntry.Executors[i];
-                        break;
-                    }
-                }
-
-                var instance = actionDescriptor.ModelTypeInfo == actionDescriptor.HandlerTypeInfo ? _model : _page;
-                result = await executor(instance, arguments);
+                // If we get here we're going to render the page so we need to initialize the context.
+                pageResult.Page = _page;
+                pageResult.ViewData = pageResult.ViewData ?? _pageContext.ViewData;
```

Comment:
This is the one part I thought was gross, we need to special case `PageResult `here. The `pageResult.ViewData` will go away in my next round a of changes.